From cd7e5a8bcd6c5629d50697973389cecc0ade89a0 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Tue, 18 Mar 2014 17:30:48 -0400
Subject: [fs] cifs: Fix check for regular file in couldbe_mf_symlink()

Message-id: <1395163848-29977-8-git-send-email-sprabhu@redhat.com>
Patchwork-id: 78383
O-Subject: [RHEL7 PATCH 7/7] [bz1077230] cifs: Fix check for regular file in couldbe_mf_symlink()
Bugzilla: 1077230
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Jeff Layton <jlayton@redhat.com>

MF Symlinks are regular files containing content in a specified format.

The function couldbe_mf_symlink() checks the mode for a set S_IFREG bit
as a test to confirm that it is a regular file. This bit is also set for
other filetypes and simply checking for this bit being set may return
false positives.

We ensure that we are actually checking for a regular file by using the
S_ISREG macro to test instead.

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
Reviewed-by: Jeff Layton <jlayton@redhat.com>
Reported-by: Neil Brown <neilb@suse.de>
Reported-by: Dan Carpenter <dan.carpenter@oracle.com>
Signed-off-by: Steve French <smfrench@gmail.com>
(cherry picked from commit a9a315d41407cd1079eb815f4adae897cc08b0d2)

Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/link.c b/fs/cifs/link.c
index f6fdec5..79fcbc1 100644
--- a/fs/cifs/link.c
+++ b/fs/cifs/link.c
@@ -185,7 +185,7 @@ format_mf_symlink(u8 *buf, unsigned int buf_len, const char *link_str)
 bool
 couldbe_mf_symlink(const struct cifs_fattr *fattr)
 {
- if (!(fattr->cf_mode & S_IFREG))
+ if (!S_ISREG(fattr->cf_mode))
   /* it's not a symlink */
   return false;
 
-- 
1.7.1